kikuchipy/indexation/similarity_metrics.py

```diff
@@ -0,0 +1,394 @@
+
+        Computes similarity or distance matrix
+        between (experimental) pattern(s) and (simulated) template(s).
+
+    greater_is_better : bool, optional
+        Whether greater values correspond to more similar images, by default True.
+        Used for choosing `n_largest` metric results in `template_match`.
+    scope : MetricScope, optional
+        Describes how `metric_func`'s input parameters is structured,
+        by default MetricScope.MANY_TO_MANY.
+    flat : bool, optional
+        Whether patterns and templates are flattened before sent to `metric_func`
+        when the similarity metric is called, by default False.
+    make_compatible_to_lower_scopes : bool, optional
+        Whether to pad patterns and templates with single dimensions
+        to match given scope, by default False.
+    dtype_out : data-type, optional
+        The data type used and returned by the metric, by default np.float32
+
+    Returns
+    -------
+    Union[SimilarityMetric, FlatSimilarityMetric]
+        A callable class instance computing a similarity matrix with signature
+        `metric(patterns,templates)`.
+    """
+    sign = 1 if greater_is_better else -1
+    if flat:
+        return FlatSimilarityMetric(
+            metric_func,
+            sign,
+            scope,
+            flat,
+            make_compatible_to_lower_scopes,
+            dtype_out,
+        )
+    else:
+        return SimilarityMetric(
+            metric_func,
+            sign,
+            scope,
+            flat,
+            make_compatible_to_lower_scopes,
+            dtype_out,
+        )
+
+
+class SimilarityMetric:
+    """Similarity metric between 2D gray-tone images."""
+
+    _P_T_NDIM_TO_SCOPE = {
+        (4, 3): MetricScope.MANY_TO_MANY,
+        (2, 3): MetricScope.ONE_TO_MANY,
+        (4, 2): MetricScope.MANY_TO_ONE,
+        (2, 2): MetricScope.ONE_TO_ONE,
+    }
+
+    _SCOPE_TO_P_T_NDMI = {
+        MetricScope.MANY_TO_MANY: (4, 3),
+        MetricScope.ONE_TO_MANY: (2, 3),
+        MetricScope.MANY_TO_ONE: (4, 2),
+        MetricScope.ONE_TO_ONE: (2, 2),
+    }
+
+    _SCOPE_TO_LOWER_SCOPES = {
+        MetricScope.MANY_TO_MANY: (
+            MetricScope.MANY_TO_ONE,
+            MetricScope.ONE_TO_MANY,
+            MetricScope.ONE_TO_ONE,
+        ),
+        MetricScope.ONE_TO_MANY: (
+            MetricScope.ONE_TO_MANY,
+            MetricScope.ONE_TO_ONE,
+        ),
+        MetricScope.ONE_TO_ONE: (),
+        MetricScope.MANY_TO_ONE: (
+            MetricScope.MANY_TO_ONE,
+            MetricScope.ONE_TO_ONE,
+        ),
+    }
+
+    def __init__(
+        self,
+        metric_func,
+        sign,
+        scope,
+        flat,
+        make_compatible_to_lower_scopes,
+        dtype_out=np.float32,
+    ):
+        self._metric_func = metric_func
+        self._make_compatible_to_lower_scopes = make_compatible_to_lower_scopes
+        self._dtype_out = dtype_out
+        self.sign = sign
```

Comment:
Is there any reason for the user to want to change any of these class attributes after initialization?

If yes, perhaps a `SimilarityMetric.greater_is_better` attribute should be added?

Reply:
The user shall in general not change the attributes. Some of the attributes will be read outside the class, that was my reasoning. I can add an underscore prefix to them all.

Reply:
I guess we decided on a wait and see approach, and leave it as it is for now.